imports/client/hooks/useCallState.ts

```diff
@@ -0,0 +1,805 @@
+        connectRef.current = callback;
+        // No need to set a callback here, since the ConnectAck record acts as a
+        // callback
+        mediasoupConnectTransport.call({
+          transportId: _id,
+          dtlsParameters: JSON.stringify(clientDtlsParameters),
+        }, (err) => {
+          if (err) {
+            console.error(`Failed to connect transport ${direction}`, err);
+          }
+        });
+      });
+      log(`setting ${direction} transport`, newTransport);
+      dispatch({
+        type: 'set-transport',
+        direction,
+        transport: newTransport,
+      });
+      return () => {
+        if (!newTransport.closed) newTransport.close();
+      };
+    } else {
+      log(`clearing ${direction} transport`);
+      dispatch({
+        type: 'set-transport',
+        direction,
+        transport: undefined,
+      });
+      return undefined;
+    }
+  }, [
+    device, hasParams, transportParams?._id, direction,
+    transportParams?.transportId, transportParams?.iceParameters, transportParams?.iceCandidates,
+    transportParams?.dtlsParameters, dispatch,
+  ]);
+
+  return connectRef;
+};
+
+function cleanupProducerMapEntry(map: Map<string, ProducerState>, trackId: string) {
+  const producerState = map.get(trackId);
+  if (producerState) {
+    // Stop the producer if present
+    if (producerState.producer) {
+      log('stopping producer for track', trackId);
+      producerState.producer.close();
+    }
+
+    // Stop the producer sub if present.
+    if (producerState.subHandle) {
+      log('stopping producer sub for track', trackId);
+      producerState.subHandle.stop();
+    }
+
+    // Drop the removed track from the producerMapRef.
+    log('producerMapRef.delete', trackId);
+    map.delete(trackId);
+  }
+}
+
+type ProducerCallback = ({ id }: { id: string }) => void;
+type ProducerState = {
+  producer: types.Producer | undefined;
+  subHandle: Meteor.SubscriptionHandle | undefined;
+  producerServerCallback: ProducerCallback | undefined;
+  kind: string | undefined;
+  rtpParameters: string | undefined;
+}
+
+type ConsumerState = {
+  consumer: types.Consumer | undefined;
+  peerId: string;
+}
+
+const useCallState = ({ huntId, puzzleId, tabId }: {
+  huntId: string,
+  puzzleId: string,
+  tabId: string,
+}): [CallState, React.Dispatch<Action>] => {
+  const [state, dispatch] = useReducer(reducer, INITIAL_STATE);
+
+  useEffect(() => {
+    // If huntId, puzzleId, or tabId change (but mostly puzzleId), reset
+    // call state.
+    return () => {
+      log('huntId/puzzleId/tabId changed, resetting call state');
+      dispatch({ type: 'reset' });
+    };
+  }, [huntId, puzzleId, tabId]);
+
+  useEffect(() => {
+    // When mediaSource (the mic capture stream) changes, stop all the tracks.
+    return () => {
+      if (state.audioState?.mediaSource) {
+        state.audioState?.mediaSource.getTracks().forEach((track) => {
+          track.stop();
+        });
+      }
```

Comment:
This will fire on the new tracks, right? Do you want to capture a reference to the old `mediaSource` in the `useEffect` before the unwind callback?

Reply:
Ah, that's probably wise, thanks.  In practice, I think the way we've written the rest of our code means `mediaSource` never transitions from truthy to truthy, so we never hit this edge case, but it's good to be resilient.